File: src/tokenizer/tokenizer.cpp
#include "tokenizer.h"

std::vector<size_t> BPETokenizer::train(const std::wstring &text, size_t vocab_size)
{
    this->text = text;
    std::vector<std::wstring> processed_text(text.size());
    size_t unique_counter = 0;
    for (size_t i = 0; i != text.size(); i++)
    {
        std::wstring unique_word(1, text[i]);
        if (iswspace(text[i]) && i > 0)
        {
            processed_text[i] = L"空";
        }
        else
        {
            processed_text[i] = unique_word;
            if (std::find(unique_words.begin(), unique_words.end(), unique_word) == unique_words.end())
            {
                unique_words.push_back(unique_word);
                vocab.try_emplace(unique_counter, unique_word);
                inverse_vocab.try_emplace(unique_word, unique_counter);
                unique_counter++;
            }
        }
    }
    unique_words.push_back(L"空");
    vocab[unique_counter] = L"空";
    inverse_vocab[L"空"] = unique_counter;
    unique_counter++;

    std::vector<size_t> token_ids(processed_text.size());
    for (size_t i = 0; i != processed_text.size(); i++)
    {
        token_ids[i] = inverse_vocab[processed_text[i]];
    }
    for (size_t new_id = unique_words.size(); new_id != unique_words.size() + vocab_size && token_ids.size() != 1; new_id++)
    {
        std::wcout << new_id << "\n";
        auto pair_id = find_freq_pair(token_ids);

        token_ids = replace_pair(token_ids, pair_id, new_id);
        bpe_merges[pair_id] = new_id;
    }

    for (auto it = bpe_merges.begin(); it != bpe_merges.end(); ++it)
    {
        std::pair<size_t, size_t> new_pair = it->first;
        size_t new_id = it->second;
        std::wstring new_token = vocab[new_pair.first] + vocab[new_pair.second];
        vocab[new_id] = new_token;
        inverse_vocab[new_token] = new_id;
    }
    return token_ids;
}

std::pair<size_t, size_t> BPETokenizer::find_freq_pair(const std::vector<size_t> &token_ids)
{
    std::vector<std::pair<size_t, size_t>> pairs(token_ids.size() - 1);
    std::map<std::pair<size_t, size_t>, size_t> pairs_counter;
    std::pair<size_t, size_t> max_pair;
    size_t max_counter = 0;
    for (size_t i = 0; i != token_ids.size() - 1; i++)
    {
        pairs[i] = std::make_pair(token_ids[i], token_ids[i + 1]);
    }
    for (size_t i = 0; i != pairs.size(); i++)
    {
        if (pairs_counter.find(pairs[i]) == pairs_counter.end())
        {
            size_t counter = 0;
            for (size_t j = 0; j != pairs.size(); j++)
            {
                if (pairs[i].first == pairs[j].first && pairs[i].second == pairs[j].second)
                {
                    counter++;
                }
            }
            pairs_counter[pairs[i]] = counter;
            if (counter > max_counter)
            {
                max_counter = counter;
                max_pair = pairs[i];
            }
        }
    }
    return max_pair;
}

std::vector<size_t> BPETokenizer::replace_pair(const std::vector<size_t> &token_ids,
                                               const std::pair<size_t, size_t> &pair_id,
                                               size_t new_id)
{
    std::vector<size_t> new_token_ids;
    for (size_t i = 0; i < token_ids.size() - 1; i++)
    {
        if (token_ids[i] == pair_id.first && token_ids[i + 1] == pair_id.second)
        {
            new_token_ids.push_back(new_id);
            i++;
        }
        else
        {
            new_token_ids.push_back(token_ids[i]);
        }
        if (i == token_ids.size() - 2)
        {
            new_token_ids.push_back(token_ids[i + 1]);
        }
    }
    return new_token_ids;
}

std::wstring BPETokenizer::decode(const std::vector<size_t> &token_ids)
{
    std::wstring decoded_string;
    for (size_t token_id : token_ids)
    {
        if (vocab.find(token_id) == vocab.end())
        {
            throw std::logic_error("decode: no such token in vocab!");
        }
        std::wstring token = vocab[token_id];
        if (token[0] == L'空')
        {
            token.erase(0, 1);
            decoded_string += L" ";
        }
        decoded_string += token;
    }
    return decoded_string;
}

std::vector<size_t> BPETokenizer::encode(const std::wstring &text)
{
    std::vector<std::wstring> tokens;
    std::wstring word;
    for (size_t i = 0; i != text.size(); i++)
    {
        if (iswspace(text[i]) && i > 0)
        {
            tokens.push_back(word);
            word.clear();
            word += L"空";
        }
        else
        {
            word += text[i];
        }
    }
    tokens.push_back(word);
    std::vector<size_t> token_ids;
    for (auto &token : tokens)
    {
        if (inverse_vocab.find(token) != inverse_vocab.end())
        {
            size_t token_id = inverse_vocab[token];
            token_ids.push_back(token_id);
        }
        else
        {
            std::vector<size_t> tokenized_word = tokenize_word(token);
            token_ids.insert(token_ids.end(), tokenized_word.begin(), tokenized_word.end());
        }
    }
    return token_ids;
}

std::vector<size_t> BPETokenizer::tokenize_word(const std::wstring &token)
{
    std::vector<size_t> token_ids(token.size());
    for (size_t i = 0; i < token.size(); i++)
    {
        std::wstring symbol = std::wstring(1, token[i]);
        if (inverse_vocab.find(symbol) == inverse_vocab.end())
        {
            throw std::logic_error("tokenize_with_bpe: no such symbol in vocab");
        }
        token_ids[i] = inverse_vocab[symbol];
    }
    bool can_merge = true;
    while (can_merge && token_ids.size() > 1)
    {
        can_merge = false;
        std::vector<size_t> new_token_ids;
        size_t i = 0;
        while (i < token_ids.size() - 1)
        {
            std::pair<size_t, size_t> token_pair(token_ids[i], token_ids[i + 1]);
            if (bpe_merges.find(token_pair) != bpe_merges.end())
            {
                size_t merged_token_id = bpe_merges[token_pair];
                new_token_ids.push_back(merged_token_id);
                i += 2;
                can_merge = true;
            }
            else
            {
                new_token_ids.push_back(token_ids[i]);
                i += 1;
            }
        }
        if (i < token_ids.size())
        {
            new_token_ids.push_back(token_ids[i]);
        }
        token_ids = new_token_ids;
    }
    return token_ids;
}

File: src/tokenizer.py
from collections import Counter, deque
from functools import lru_cache
import json
import sys


class BPETokenizerSimple:
    def __init__(self):
        self.vocab = {}
        self.inverse_vocab = {}
        self.bpe_merges = {}

    def train(self, text, vocab_size):
        processed_text = []
        for i, char in enumerate(text):
            if char.isspace() and i != 0:
                processed_text.append("Ġ")
            if char != " ":
                processed_text.append(char)
        processed_text = "".join(processed_text)

        unique_chars = [char for char in sorted(set(processed_text))]
        unique_chars.append('Ġ')

        self.vocab = {i: char for i, char in enumerate(unique_chars)}
        self.inverse_vocab = {char: i for i, char in self.vocab.items()}

        token_ids = [self.inverse_vocab[char] for char in processed_text]

        for new_id in range(len(self.vocab), vocab_size):
            print(new_id)
            pair_id = self.find_freq_pair(token_ids)
            token_ids = self.replace_pair(token_ids, pair_id, new_id)
            self.bpe_merges[pair_id] = new_id

        for (p0, p1), new_id in self.bpe_merges.items():
            merged_token = self.vocab[p0] + self.vocab[p1]
            self.vocab[new_id] = merged_token
            self.inverse_vocab[merged_token] = new_id

    def encode(self, text):
        tokens = []
        words = text.replace("\n", " \n ").split()

        for i, word in enumerate(words):
            if i > 0 and not word.startswith("\n"):
                tokens.append("Ġ" + word)
            else:
                tokens.append(word)

        token_ids = []
        for token in tokens:
            if token in self.inverse_vocab:
                token_id = self.inverse_vocab[token]
                token_ids.append(token_id)
            else:
                sub_token_ids = self.tokenize_with_bpe(token)
                token_ids.extend(sub_token_ids)

        return token_ids

    def tokenize_with_bpe(self, token):
        token_ids = [self.inverse_vocab.get(char, None) for char in token]
        if None in token_ids:
            missing_chars = [char for char, tid in zip(token, token_ids) if tid is None]
            raise ValueError(f"Characters not found in vocab: {missing_chars}")

        can_merge = True
        while can_merge and len(token_ids) > 1:
            can_merge = False
            new_tokens = []
            i = 0
            while i < len(token_ids) - 1:
                pair = (token_ids[i], token_ids[i + 1])
                if pair in self.bpe_merges:
                    merged_token_id = self.bpe_merges[pair]
                    new_tokens.append(merged_token_id)
                    i += 2  
                    can_merge = True
                else:
                    new_tokens.append(token_ids[i])
                    i += 1
            if i < len(token_ids):
                new_tokens.append(token_ids[i])
            token_ids = new_tokens

        return token_ids

    def decode(self, token_ids):
        decoded_string = ""
        for token_id in token_ids:
            if token_id not in self.vocab:
                raise ValueError(f"Token ID {token_id} not found in vocab.")
            token = self.vocab[token_id]
            if token.startswith("Ġ"):
                # Replace 'Ġ' with a space
                decoded_string += " " + token[1:]
            else:
                decoded_string += token
        return decoded_string

    def save_vocab_and_merges(self, vocab_path, bpe_merges_path):
        # Save vocabulary
        with open(vocab_path, "w", encoding="utf-8") as file:
            json.dump({k: v for k, v in self.vocab.items()}, file, ensure_ascii=False, indent=2)

        # Save BPE merges as a list of dictionaries
        with open(bpe_merges_path, "w", encoding="utf-8") as file:
            merges_list = [{"pair": list(pair), "new_id": new_id}
                           for pair, new_id in self.bpe_merges.items()]
            json.dump(merges_list, file, ensure_ascii=False, indent=2)

    def load_vocab_and_merges(self, vocab_path, bpe_merges_path):
        with open(vocab_path, "r", encoding="utf-8") as file:
            loaded_vocab = json.load(file)
            self.vocab = {int(k): v for k, v in loaded_vocab.items()}
            self.inverse_vocab = {v: int(k) for k, v in loaded_vocab.items()}

        with open(bpe_merges_path, "r", encoding="utf-8") as file:
            merges_list = json.load(file)
            for merge in merges_list:
                pair = tuple(merge['pair'])
                new_id = merge['new_id']
                self.bpe_merges[pair] = new_id

    def find_freq_pair(token_ids):
        pairs = Counter(zip(token_ids, token_ids[1:]))
        return max(pairs.items(), key=lambda x: x[1])[0]

    def replace_pair(token_ids, pair_id, new_id):
        dq = deque(token_ids)
        replaced = []

        while dq:
            current = dq.popleft()
            if dq and (current, dq[0]) == pair_id:
                replaced.append(new_id)
                dq.popleft()
            else:
                replaced.append(current)

        return replaced